File: csharp/139_-_Word_Break.cs
/**
 * 139. Word Break - Medium
 * Given a string s and a dictionary of strings wordDict, return true if s can
 * be segmented into a space-separated sequence of one or more dictionary words.
 * Note that the same word in the dictionary may be reused multiple times in the
 * segmentation.
 */

public class Solution {
    public bool WordBreak(string s, IList<string> wordDict) {
        bool[] memo = new bool[s.Length + 1];
        Array.Fill(memo, false);
        memo[s.Length] = true;
        
        for (int i = s.Length - 1; i >= 0; --i) {
            foreach (string word in wordDict) {
                if (i + word.Length <= s.Length && s.Substring(i, word.Length) == word) {
                    memo[i] = memo[i + word.Length];
                }
                if (memo[i] == true) {
                    break;
                }
            }
        }
        
        return memo[0];
    }
}

File: javascript/139_-_Word_Break.js
/**
 * 139. Word Break - Medium
 * Given a string s and a dictionary of strings wordDict, return true if s can
 * be segmented into a space-separated sequence of one or more dictionary words.
 * Note that the same word in the dictionary may be reused multiple times in the
 * segmentation.
 */

/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
var wordBreak = function(s, wordDict) {
    const memo = new Array(s.length + 1);
    memo.fill(false);
    memo[s.length] = true;
    
    for (let i = s.length - 1; i >= 0; --i) {
        for (const word of wordDict) {
            if (i + word.length <= s.length && word === s.substring(i, i + word.length)) {
                memo[i] = memo[i + word.length];
            }
            if (memo[i]) {
                break;
            }
        }
    }
    
    return memo[0];
};